pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py

```diff
@@ -0,0 +1,545 @@
+"""Tests for the data migration in `filename`."""
+import logging
+from collections import ChainMap, namedtuple
+from datetime import timedelta
+from itertools import count
+from typing import Dict, Iterable, Type, Union
+
+from django.db.models import Q
+from django.forms.models import model_to_dict
+from django.utils import timezone
+
+from pydis_site.apps.api.models import Infraction, User
+from .base import MigrationsTestCase
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+InfractionHistory = namedtuple('InfractionHistory', ("user_id", "infraction_history"))
+
+
+class InfractionFactory:
+    """Factory that creates infractions for a User instance."""
+
+    infraction_id = count(1)
+    user_id = count(1)
+    default_values = {
+        'active': True,
+        'expires_at': None,
+        'hidden': False,
+    }
+
+    @classmethod
+    def create(
+        cls,
+        actor: User,
+        infractions: Iterable[Dict[str, Union[str, int, bool]]],
+        infraction_model: Type[Infraction] = Infraction,
+        user_model: Type[User] = User,
+    ) -> InfractionHistory:
+        """
+        Creates `infractions` for the `user` with the given `actor`.
+
+        The `infractions` dictionary can contain the following fields:
+         - `type` (required)
+         - `active` (default: True)
+         - `expires_at` (default: None; i.e, permanent)
+         - `hidden` (default: False).
+
+         The parameters `infraction_model` and `user_model` can be used to pass in an instance of
+         both model classes from a different migration/project state.
+        """
+        user_id = next(cls.user_id)
+        user = user_model.objects.create(
+            id=user_id,
+            name=f"Infracted user {user_id}",
+            discriminator=user_id,
+            avatar_hash=None,
+        )
+        infraction_history = []
+
+        for infraction in infractions:
+            infraction = dict(infraction)
+            infraction["id"] = next(cls.infraction_id)
+            infraction = ChainMap(infraction, cls.default_values)
+            new_infraction = infraction_model.objects.create(
+                user=user,
+                actor=actor,
+                type=infraction["type"],
+                reason=f"`{infraction['type']}` infraction (ID: {infraction['id']} of {user}",
+                active=infraction['active'],
+                hidden=infraction['hidden'],
+                expires_at=infraction['expires_at'],
+            )
+            infraction_history.append(new_infraction)
+
+        return InfractionHistory(user_id=user_id, infraction_history=infraction_history)
+
+
+class InfractionFactoryTests(MigrationsTestCase):
```

Comment:
Why is this a `MigrationsTestCase`? The tests for the factory aren't concerned with what values are allowed for fields but rather just that if a given value is equal to the resulting value. Is it just so future migrations don't break this test (e.g. in the future a check constraint for a kick would be added that it must always be hidden, so line 105 would fail)?

Reply:
> Is it just so future migrations don't break this test (e.g. in the future a check constraint for a kick would be added that it must always be hidden, so line 105 would fail)?

Yes, it's to pin these tests to the point in the migration history we are testing. For our non-migration tests, we always update the tests to reflect the current (i.e., latest) states of the models, but that would not make sense for something that is used specifically to generate data according to an older model.

In this case, I want to use `InfractionFactory` to specifically generate data that would violate the UniqueTogether constraint introduced in a later migration.

Reply:
> In this case, I want to use InfractionFactory to specifically generate data that would violate the UniqueTogether constraint introduced in a later migration.

That makes perfect sense, I did not catch this use myself.